src/server/system_services/system_server.js

```diff
@@ -1309,6 +1309,277 @@ async function _get_endpoint_groups() {
+    const entity_info = get_entity_info(entity, entity_type);
+    if (!entity_info) throw new Error(`_disable_master_key: entity doesn't exist ${entity}`);
+
+    if (!entity_info.master_key_id) {
+        throw new Error(`_disable_master_key: master key of entity: ${entity} is undefined - can not disable master key`);
+    }
+    if (entity_info.master_key_id.disabled === true) {
+        throw new Error(`_disable_master_key: master key of entity: ${entity} is already disabled`);
+    }
+    await update_insert_master_key({ op: 'update', _id: entity_info.master_key_id._id, update: {disabled: true}});
+    system_store.master_key_manager.set_m_key_disabled_val(entity_info.master_key_id._id, true);
+
+    if (entity_type === 'ACCOUNT' || entity_type === 'SYSTEM') {
+        await _execute_func_on_sub_entities(entity_info, entity_type, entity_info.master_key_id._id, undefined, 'DISABLE');
+    }
+}
+
+/**
+ *
+ * ENABLE MASTER KEY
+ *
+ */
+async function enable_master_key(req) {
+    const { entity, entity_type } = req.rpc_params; // System Name, Bucket Name, Account Email
+    await _enable_master_key(entity, entity_type);
+}
+
+// enabling account - need to encrypt all the access keys.
+// enabling system, we don't enable the lower sub trees
+// enabling bucket, the map builder bg worker will encrypt the cipher chunks and set the bucket master key to the chunk. 
+async function _enable_master_key(entity, entity_type) {
+    const entity_info = get_entity_info(entity, entity_type);
+    if (!entity_info) throw new Error(`_enable_master_key: entity doesn't exist ${entity}`);
+
+    if (!entity_info.master_key_id) {
+        throw new Error(`_enable_master_key: master key of entity: ${entity} is undefined - can not enable master key`);
+    }
+    if (entity_info.master_key_id.disabled === false) {
+        throw new Error(`_enable_master_key: master key of entity: ${entity} is already enabled`);
+    }
+    await update_insert_master_key({ op: 'update', _id: entity_info.master_key_id._id, update: { disabled: false }});
+    system_store.master_key_manager.set_m_key_disabled_val(entity_info.master_key_id._id, false);
+
+    if (entity_type === 'ACCOUNT') {
+        await _execute_func_on_sub_entities(entity_info, entity_type, undefined, entity_info.master_key_id._id, 'ENABLE');
+    }
+}
+
+async function _execute_func_on_sub_entities(entity_info, entity_type, old_master_key_id, new_master_key_id, m_key_function) {
+    await system_store.load();
+    switch (entity_type) {
+        case 'ACCOUNT': {
+            const access_keys = _reencrypt_or_decrypt_creds(entity_info.access_keys, new_master_key_id, m_key_function);
+            const sync_credentials_cache = _reencrypt_or_decrypt_creds(entity_info.sync_credentials_cache,
+                new_master_key_id, m_key_function);
+            const pools_ns_resource_updates = get_pools_and_ns_resources_changes(entity_info.sync_credentials_cache, entity_info._id);
```

Comment:
You place the old keys instead of the new keys.
But you change the values inside the function for the parameters (this is why it works).
Really confusing to understand and read.

Reply:
fixed